pkg/storage/tree/tree_profile.go

```diff
@@ -0,0 +1,181 @@
+)
+
+type ProfileReader struct {
+	sampleTypesFilter func(string) bool
+	cache             labelsCache
+}
+
+func NewProfileReader() *ProfileReader { return new(ProfileReader) }
+
+func (r *ProfileReader) SampleTypeFilter(fn func(string) bool) *ProfileReader {
+	r.sampleTypesFilter = fn
+	return r
+}
+
+func (r *ProfileReader) Reset() {
+	r.cache = make(labelsCache)
+}
+
+func (r *ProfileReader) Load(st int, labels Labels) (*Tree, bool) {
+	e, ok := r.cache.get(st, labels.Hash())
+	if !ok {
+		return nil, false
+	}
+	return e.Tree, true
+}
+
+func (r *ProfileReader) Read(x *Profile, fn func(vt *ValueType, l Labels, t *Tree) (keep bool, err error)) error {
+	sampleTypes := make([]int, 0, len(x.SampleType))
+	for i, s := range x.SampleType {
+		if r.sampleTypesFilter != nil && !r.sampleTypesFilter(x.StringTable[s.Type]) {
+			continue
+		}
+		sampleTypes = append(sampleTypes, i)
+	}
+	if len(sampleTypes) == 0 {
+		return nil
+	}
+	c := make(labelsCache)
+	readTrees(x, sampleTypes, c, NewFinder(x))
+	for sampleType, entries := range c {
+		for h, e := range entries {
+			keep, err := fn(x.SampleType[sampleType], e.Labels, e.Tree)
+			if err != nil {
+				return err
+			}
+			if !keep {
+				c.remove(sampleType, h)
+			}
+		}
+	}
+	r.cache = c
+	return nil
+}
+
+func (x *Profile) ResolveLabels(l Labels) map[string]string {
+	m := make(map[string]string)
+	for _, label := range l {
+		if label.Str != 0 {
+			m[x.StringTable[label.Key]] = x.StringTable[label.Str]
+		}
+	}
+	return m
+}
+
+type Labels []*Label
+
+func (l Labels) Hash() uint64 {
+	h := xxhash.New()
+	t := make([]byte, 16)
+	for _, x := range l {
+		if x.Str == 0 {
+			continue
+		}
+		binary.LittleEndian.PutUint64(t[0:8], uint64(x.Key))
+		binary.LittleEndian.PutUint64(t[8:16], uint64(x.Str))
+		_, _ = h.Write(t)
+	}
+	return h.Sum64()
+}
+
+func readTrees(x *Profile, sampleTypes []int, c labelsCache, f Finder) {
+	stack := make([][]byte, 0, 16)
+	for _, s := range x.Sample {
+		for i := len(s.LocationId) - 1; i >= 0; i-- {
+			// Resolve stack.
+			loc, ok := f.FindLocation(s.LocationId[i])
+			if !ok {
+				continue
+			}
+			// Multiple line indicates this location has inlined functions,
+			// where the last entry represents the caller into which the
+			// preceding entries were inlined.
+			//
+			// E.g., if memcpy() is inlined into printf:
+			//    line[0].function_name == "memcpy"
+			//    line[1].function_name == "printf"
+			//
+			// Therefore iteration goes in reverse order.
+			for j := len(loc.Line) - 1; j >= 0; j-- {
+				fn, ok := f.FindFunction(loc.Line[j].FunctionId)
+				if !ok {
+					continue
+				}
+				stack = append(stack, unsafeStrToSlice(x.StringTable[fn.Name]))
+			}
+		}
+		// Insert tree nodes.
+		for _, sampleType := range sampleTypes {
+			if v := uint64(s.Value[sampleType]); v != 0 {
+				c.getOrCreate(sampleType, s.Label).InsertStack(stack, v)
+			}
+		}
+		stack = stack[:0]
+	}
+}
+
+func unsafeStrToSlice(s string) []byte {
```

Comment:
This is a bit scary, I think it may be possible to remove the copy without relying on `unsafe` with a bit of extra work in tree insertion (e.g. adding some `func (n *treeNode) insertString` that iterates over the bytes of the string (which should then be [optimized by the compiler](https://github.com/golang/go/wiki/CompilerOptimizations#range-over-bytes)).
Just thinking out loud here.

Reply:
Thank you for the suggestion - this an awesome idea.

I think the same - instead of standard binary search (which does not even seem to be inlined) we could implement something more specific for our case. Regarding safety: although this is a widely adopted technique, I would rather not use it. I put it as it is the easiest way to avoid allocation here.

Reply:
Coincidentally I ended up needing something like this for the flamebearer to tree conversion, so I have added some initial support for it in https://github.com/pyroscope-io/pyroscope/pull/652/commits/603201fed214b294b8737614d1bdb77b2d27679a.
I went with the standard binary search to validate it quickly, but I'll change that (and also add tests). Let me know if you have further insights :)

Reply:
Unsurprisingly, the implementation had several bugs :sweat_smile:
Changed to use custom binary search and added tests in https://github.com/pyroscope-io/pyroscope/pull/652/commits/672d2156fd92aae5d88310e8f28e9898a96aa63c
A further improvement would be to use BSTs instead of slices for tree nodes, but that may be overkill for this thread, we can discuss that elsewhere.